### PR TITLE
RxMobileWalletAdapter Convenience APIs

### PR DESCRIPTION
**Pull request by olivierg13**
Probably not merging this PR as it's modifying/breaking the `fakedapp` example.
Just a way to propose and discuss around the Rx convenience APIs.

### CODE REVIEW

File: android/clientlib-rxjava/src/main/java/com/solana/mobilewalletadapter/clientlib/ActivityResultSender.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.solana.mobilewalletadapter.clientlib;
+
+import android.content.Intent;
+
+/**
+ * Used to abstract away which API is used for starting/launching an Activity for result, depending
+ * on API level, etc...
+ */
+public interface ActivityResultSender {
```
Comment (creativedrewy) [inline android/clientlib-rxjava/src/main/java/com/solana/mobilewalletadapter/clientlib/ActivityResultSender.java:9]: No `I`. My inner C# developer is crying.

File: android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt [hunk 2]
```diff
@@ -33,11 +36,57 @@ class MainViewModel : ViewModel() {
     private val _uiState = MutableStateFlow(UiState())
     val uiState = _uiState.asStateFlow()
 
-    private val mobileWalletAdapterClientSem = Semaphore(1) // allow only a single MWA connection at a time
+    private val compositeDisposable = CompositeDisposable()
 
-    suspend fun authorize(sender: StartActivityForResultSender) {
-        localAssociateAndExecute(sender) { client ->
-            doAuthorize(client)
+    private val mobileWalletAdapterClientSem =
```
Comment (creativedrewy) [inline android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt:41]: Wait, do we want to make the fakedapp use Rx? Or should we get a few different screens+ViewModels going here?

**Review by creativedrewy — COMMENTED** [review]
I like it, but let's sync around what we want to do with MainViewModel

File: android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt [hunk 3]
```diff
@@ -310,10 +375,26 @@ class MainViewModel : ViewModel() {
...
-                        ProtocolContract.ERROR_TOO_MANY_PAYLOADS -> Log.e(TAG, "Too many payloads to sign", cause)
+                        ProtocolContract.ERROR_REAUTHORIZE -> Log.e(
+                            TAG,
+                            "Reauthorization required",
+                            cause
+                        )
+                        ProtocolContract.ERROR_AUTHORIZATION_FAILED -> Log.e(
+                            TAG,
+                            "Auth token invalid",
+                            cause
```
Comment (creativedrewy) [inline android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt:386]: Ugh, can we pleas not have our code auto format this way? It's so many newlines for very limited readablity.

File: android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt (hunk 2 above)
Comment (olivierg13) [reply android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt:41]: Yeah definitely, different screens is needed in the future.
As mentioned in the description,  I won't merge this PR and it's mainly for discussion.

File: android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt (hunk 3 above)
Comment (olivierg13) [reply android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt:386]: Ugh yeah 😬 I thought I had reverted these auto-format changes. Will try to revert them back from the source.

File: android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt (hunk 2 above)
Comment (creativedrewy) [reply android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt:41]: Ooooh, sorry! Forgot that.

File: android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt (hunk 3 above)
Comment (olivierg13) [reply android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt]: Should be good now

**Review by creativedrewy — APPROVED** [review]
This PR is ready when you are

File: android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt [hunk 4]
```diff
@@ -33,12 +36,61 @@ class MainViewModel : ViewModel() {
...
+
+    override fun onCleared() {
+        compositeDisposable.dispose()
+        super.onCleared()
+    }
+
+    fun authorize(sender: ActivityResultSender) {
+        RxMobileWalletAdapter(Scenario.DEFAULT_CLIENT_TIMEOUT_MS, null)
+            .transact(sender)
+            .subscribe { rxMobileWalletAdapterClient ->
```
[the remark this answers is not in the document]
Comment (olivierg13) [reply android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt]: Oups sorry I just deleted the examples 😬 

Essentially within the first subscribe you will have to do:
```
  authorize()
    .subscribe/map {
      // result of the authorize (authToken)
      signAndSendTransaction()
        .subscribe {
          // result of the signAndSend
        }
    }
```

Rx is very much indented compared to coroutines for sure (especially when you need the result from a previous Observable).

[the remark this answers is not in the document]
Comment (creativedrewy) [reply android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt]: Yeah the indentation is kinda obnoxious, but I think most uses won't be more than a couple deep.

**Pull request merged**